datastore/longhorn.go

```diff
@@ -1109,6 +1112,29 @@ func GetNewCurrentEngineAndExtras(v *longhorn.Volume, es map[string]*longhorn.En
 		}
 	}
 
+	// Corner case:
+	// During a migration confirmation, we make the following requests to the API server:
+	// 1. Delete the active engine (guaranteed to succeed or we don't do anything else).
+	// 2. Switch the new current replicas to active (might fail, preventing everything below).
+	// 3. Set the new current engine to active (might fail, preventing everything below).
+	// 4. Set the volume.Spec.CurrentNodeID (might fail).
+	// So we might delete the active engine, fail to set the new current engine to active, and fail to set the
+	// volume.Spec.CurrentNodeID. Then, the volume attachment controller might try to do a full detachment, setting
+	// volume.Spec.NodeID == "". At this point, there is no engine that can ever be considered active again by the rules
+	// above, but we want to use the new current engine. It's engine.Spec.NodeID ==
+	// volume.Status.CurrentMigrationNodeID.
```

Comment:
> So we might delete the active engine, fail to set the volume.Spec.CurrentNodeID.

> At this point, there is no engine that can ever be considered active again by the rules above, but we want to use the new current engine

Will Longhorn retry marking this new engine as active later?

Reply:
No. We will only set an engine to active if it is returned from `GetNewCurrentEngineAndExtras`.

https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/controller/volume_controller.go#L3849

This function disallows engines with deletion timestamps.

https://github.com/longhorn/longhorn-manager/blob/e5701524c9610cb780ef06960d8ede62f25fa7ec/datastore/longhorn.go#L1101-L1103

Reply:
Discussed this with @shuo-wu offline. I actually misunderstood his question. The answer is yes. Steps 2-4 modify the associated objects mid-reconcile, but don't attempt to make changes on the API server until the end. If some error occurs (e.g. a conflict), we will rereconcile and repeat the actions.